### ISSUE TITLE
-t and -d don't work as expected

### ISSUE THREAD

**Issue by sandrotosi** [wireservice/csvkit#666]
Hello,
this has been reported in Debian as http://bugs.debian.org/824252

```

The -t (--tabs) and -d (--delimiter) options only change the
input delimiter, while the output delimiter is hard-coded to
be a comma.

There doesn't appear to be any way to change the output delimiter.

Using tr or similar to work around this is unreliable - there may be
actual commas in any of the tab-separated fields.  It also defeats the
purpose of using specialised tools like csvcut.

e.g. see example usage of csvcut and comments at
http://unix.stackexchange.com/a/186354/7696


The output delimiter should either be the same as the input delimiter
or there should be extra options (perhaps -T and -D) to set it
(similar to the way awk's FS and OFS variables work)
```

**Comment by jpmckinney** [comment]
Not a bug. You must pipe the output of a command into `csvformat` to change the output format.

**Comment by craig-sanders** [comment]
That misses the point (and the problem) entirely.

Here's the real issue, quoted from my original bug report.

 > Using tr or similar to work around this is unreliable - there may be
 > actual commas in any of the tab-separated fields.

(I made a mistake there. I meant to say "any of the comma-separated fields", not "tab-separated").

If there are commas inside the fields, in the actual data, it is **impossible** for any other program that post-processes csvkit's output to distinguish between a comma within a field from a comma that separates a field.  That includes csvformat as well as tr, awk, sed, or any other tool.

**THAT** is why being able to specify the output field separator is essential - so you can choose a delimiter which **ISN'T** in the data.

If csvkit can't do that, then it's not fit for purpose.

This sort of data vs delimiter problem is very common.

If you can choose the delimiter, then it's a trivial problem to solve, easily scripted.

If you can't choose, then the **only** way to be certain that your output isn't going to be garbled by an extra comma is to visually inspect it and edit it by hand - it is a task that requires human-level intelligence and great patience and attention to detail (or at least AI parsing of the data)

**Comment by jpmckinney** [comment]
Create a pipe-delimited file, with one column header containing a comma:

```
echo 'id|a,comma|name
1|2|3
1|4|6' > test.csv
```

Run any csvkit command with the `-d` option to set the delimiter to `|`, for example: `csvgrep -d '|' -c 3 -m 3 test.csv`. The output is comma-delimited, and properly quotes the column header that contains a comma:

```
id,"a,comma",name
1,2,3
```

You can pipe the output to additional csvkit commands, without having to set the `-d` option on subsequent commands, for example: `csvgrep -d '|' -c 3 -m 3 test.csv | csvcut -c 1,2`. The output is:

```
id,"a,comma"
1,2
```

If you want to go back to having pipe-delimited file, use csvformat, for example: `csvgrep -d '|' -c 3 -m 3 test.csv | csvcut -c 1,2 | csvformat -D '|'`. The output is:

```
id|a,comma
1|2
```

**Comment by jpmckinney** [comment]
> it is impossible for any other program that post-processes csvkit's output to distinguish between a comma within a field and a comma that separates fields

It is possible because CSV writers add quotes around field values that contain delimiters.

**Comment by craig-sanders** [comment]
yes, it's very easy when you have complete control over the input data, especially when you can make up an easy example.  I know how to deal with nice, easy data.  I also know how to work with data that isn't so convenient.  perfectly formatted data is pretty much only ever seen in examples.

real world data is generally a lot uglier and messier than that, including such horrors as unquoted comma-containing fields and unmatched quote characters and abominations far worse than those.  THAT is when (and why) you need to be able to specify the output delimiter.  and, correct me if i'm wrong, but isn't that part of the intended purpose of csvkit, to massage real world data into shape for some real world purpose?



some CSV writers add quotes, many don't.  More importantly, some CSV readers interpret quotes but many don't.  and sometimes a simple pipe to awk or cut or sed or whatever is the right solution, rather than spending an hour or so writing a python or perl or whatever script to parse properly formatted CSV correctly.

**Comment by jpmckinney** [comment]
Can you provide an example that we can work with?

**Comment by craig-sanders** [comment]
Not at the moment.  I could make up an example but it wouldn't be much use, because it would only cover a tiny subset of the malformations possible in real data and would have at most a dozen or so lines rather than millions, so would seem to be a trivial problem easily fixed by hand.

I've already mentioned two of the more common problems (commas in text fields, unmatched quotes).  Others include commas in numeric fields (e.g. "1,048,576" or locales that use commas as the decimal separator), carriage returns and/or line feeds in unquoted text fields, and more.  It's difficult to remember or even think of all the weird data problems I've encountered over the decades that i've been doing this kind of data mangling - every data file or data source is different, and you need to be adaptable, and have flexible tools, to deal with them.

As a rule of thumb, the easiest and most time, RAM, & CPU efficient method is to first search the input data for a character that **doesn't** appear anywhere in the input and then use that as the output delimiter - if your data is full of commas, spaces and tabs, use colons.  If it has colons, use semi-colons or commas or tabs.  or equals, or underscore. or anything that doesn't appear in the input.  Sometimes you have to choose something that only appears rarely in the input, first transforming it to something else.

spaces in space-delimited text fields are also extremely common, but not all that relevant to csvkit's output (but can cause problems for csvkit's input processing).

These issues are often the reason **why** you're processing the file in the first place - to clean it up and normalise it because it's full of horrors like this and someone, perhaps you, needs to import it into some other program or a database or just analyse it and produce a summary.

**Comment by jpmckinney** [comment]
I'm not clear on how setting an output delimiter using an option in every csvkit command changes the situation compared to piping that output to `csvformat` and using the `-D` option to set the output delimiter at that point. For the examples of:

* unquoted comma-containing fields
* unmatched quote characters
* carriage returns and/or line feeds in unquoted text fields
* spaces in space-delimited text fields

csvkit cannot guess which of the alternative interpretations of the data is correct when reading that input. If there is a field value containing an unquoted comma, csvkit can't determine whether that row has one more column than other rows, or whether one of the fields contains an unquoted comma - and it definitely can't determine which of those fields contains that comma. Each of the examples above produces alternative interpretations, either of which may be correct. In such cases, you need a human to make the decision, and either code those decisions into a custom script, or correct the file manually. csvkit just reads CSVs using the rules of Python's csv module, while allowing for some of the [variations](https://docs.python.org/2/library/csv.html#dialects-and-formatting-parameters) that Python's csv module supports.

Going back to the unquoted comma-containing field example, I don't see how setting the output delimiter helps the situation. If you have a row `country,1,234,567,2013` - that should be read as `country` has a population of `1,234,567` in the year `2013` - running `csvcut -c 1-5 -D =` (using the proposed output delimiter option `-D`) will read it as five values, and will output `country=1=234=567=2013`. Running `csvcut -c 1-2 -D =` (because you only want the country and population) will output `country=1`, which isn't what you want, but it is the best csvkit can do. But in either case, having the `-D` option on `csvcut` instead of `csvformat` doesn't change the output...

**Comment by craig-sanders** [comment]
there may not have been a comma separator in the original input, it may have been space or tab or colon or whatever delimited.  with the -t and -d options, csvkit can read such files but forcing all output to be transformed to comma-separated makes it impossible to use csvkit's useful capabilities with such data.

which is why my initial bug report said that the output delimiter should either be the same as the input delimiter, or (preferably) it should be possible to specify the output delimiter at run time.

if you want to impose an arbitrary and pointless limitation on your software, well, it's your software and you can do what you like with it.   but the fix is, or should be trivial - and note that CSV is only a subset of DSV (see https://en.wikipedia.org/wiki/Delimiter-separated_values and e.g. http://python-dsv.sourceforge.net/)

**Comment by jpmckinney** [comment]
I still don't understand how "forcing all output to be transformed to comma-separated makes it impossible to use csvkit's useful capabilities with such data". You can just use `csvformat` to change it back. I haven't yet seen an explicit counterexample.

**Comment by onyxfish** [comment]
@craig-sanders I would appreciate it if you would comport yourself with a bit more respect for those of us who spend our time building this software. csvkit has existed for over 5 years, so you're presumption that we haven't thought this through is, frankly, insulting. Your passive aggressive use of bold text is doesn't help either.

The decision to standardize output formats is neither arbitrary, nor pointless. It is a [documented](https://csvkit.readthedocs.io/en/1.0.1/contributing.html#principles-of-development) design decision in csvkit—a decision that has served us very well. I've never seen a CSV reader, no matter how badly implemented, that doesn't support quoted fields. If it doesn't then it simply isn't a CSV reader by any useful definition.

By standardizing on output we put more well-formatted data into the world. If you really need a custom output format you can easily accommodate that in the ways @jpmckinney has described (namely `csvformat`), but I'm not going to encourage anybody to do that, nor screw up the user experience of the tools, by making it a default.

In summary, we've thought a lot about this and it works just fine the way it is.

**Comment by craig-sanders** [comment]
@jpmckinney you can't just change it back because once it has been output, you have lost (thrown away) part of the information required to make the necessary decisions - any subsequent program in the pipeline only sees the output of the most recent program, it has no access to the original input.  like any other tool, csvformat can only work with what it has, not what was discarded.

@onyxfish  wow.  bold text is "passive aggressive" now, is it?  I thought it was for emphasis, or to highlight particular words or phrases so that they stand out - which is why github indicates it in the editor with a B for bold rather than P for passive-aggressive?    how about you concentrate on the technical issues instead of dragging this down to the level of personal abuse?

as i said before, it's your software, so you can do whatever you want with it.  but don't try to pretend that it's generally useful beyond the deliberately narrow scope you're limiting it to.  and that's the only reason I care at all about it - csvkit has some capabilities that would be quite useful as a general-purpose data processing tool, saving much time and effort in rewriting boiler-plate one-shot awk or perl or python scripts.  if the only thing it can output is comma-separated, then that usefulness is merely potential, not actual - which is why my initial enthusiasm on discovering the software was greatly diminished when i realised it was crippled in this way.

**Comment by jpmckinney** [comment]
@craig-sanders It's standard for a bug report to include directions for how to reproduce the bug. You need to do that to demonstrate what you're stating. I don't think any data is being discarded that can't be reinstated. There's no point continuing this discussion without an example.

Also, I agree with @onyxfish that your choices of words and phrasing are not helping you. Providing an example input that produces the consequences you describe would be a cooperative next step.
